runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -1330,6 +1331,14 @@ private void onNewEventTimer(TimerData newTimer) {
       }
     }
 
+    /** Holds the watermark when there is an sdf timer. */
+    private void onNewSdfTimer(TimerData newTimer) {
+      Preconditions.checkState(
+          StateAndTimerBundleCheckpointHandler.isSdfTimer(newTimer.getTimerId()));
+      Preconditions.checkState(timerUsesOutputTimestamp(newTimer));
+      keyedStateInternals.addWatermarkHoldUsage(newTimer.getOutputTimestamp());
+    }
```

Comment:
Wouldn't it make sense to integrate this check with the `timerUsesOutputTimestamp` method?

Reply:
We could rename the mentioned method to `onFiredTimer` and include the checks for output watermark holds in there.

Reply:
Sounds good.

Reply:
Sorry I want to revisit the idea of having `onFiredTimer` here. I think it's a good idea to have `onFiredTimer` for firing timers. But the function `onNewSdfTimer` and `onNewEventTimer` is about to set watermark hold when registering timers. Different from event timer, an SDF timer must have the output timestamp for controlling watermark hold. It's important for SDF execution. That's why we have a check instead of an if block here.